File: code/19.go
/*
 * @Author: lmio
 * @Date: 2023-02-05 21:43:03
 * @LastEditTime: 2023-03-23 23:11:57
 * @FilePath: /leetcode/code/19.go
 * @Description:19.删除链表的倒数第N个结点
 */
package code

import . "leetcode/code_struct"

func RemoveNthFromEnd(head *ListNode, n int) *ListNode {
	right := head
	left := &ListNode{Val: 0, Next: head}
	for i := 0; i < n ; i++ {
			right = right.Next
	}
	
	for right != nil {
		right = right.Next
		left = left.Next
	}
	if left.Next == head {
		head = head.Next
	}else {
		left.Next = left.Next.Next
	}
	return head
}

File: code/19.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-07-08 11:05:40
LastEditors: lmio 2091319361@qq.com
Description: 19. 删除链表的倒数第 N 个结点
'''
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def removeNthFromEnd(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    dummy = ListNode()
    dummy.next = head
    prev = dummy
    cur = head
    for _ in range(n):
        cur = cur.next
    while cur:
        cur = cur.next
        prev = prev.next
    prev.next = prev.next.next
    return dummy.next